service/sharddistributor/leader/namespace/manager.go

```diff
@@ -0,0 +1,172 @@
+package namespace
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/fx"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/service/sharddistributor/config"
+	"github.com/uber/cadence/service/sharddistributor/leader/election"
+	"github.com/uber/cadence/service/sharddistributor/leader/process"
+)
+
+// Module provides namespace manager component for an fx app.
+var Module = fx.Module(
+	"namespace-manager",
+	fx.Provide(
+		NewManager,
+	),
+	// Force use manager to get it into fx.
+	fx.Invoke(func(Manager) {}),
+)
+
+// Manager handles namespace distribution and leader election
+type Manager interface {
+	Start(ctx context.Context) error
+	Stop(ctx context.Context) error
+}
+
+type namespaceManager struct {
+	cfg              config.LeaderElection
+	logger           log.Logger
+	electionFactory  election.Factory
+	processorFactory process.Factory
+	namespaces       map[string]*namespaceHandler
+	mu               sync.RWMutex
+	ctx              context.Context
+	cancel           context.CancelFunc
+}
+
+type namespaceHandler struct {
+	logger       log.Logger
+	elector      election.Elector
+	processor    process.Processor
+	cancel       context.CancelFunc
+	namespaceCfg config.Namespace
+	cleanupWg    sync.WaitGroup
+}
+
+type ManagerParams struct {
+	fx.In
+
+	Cfg              config.LeaderElection
+	Logger           log.Logger
+	ElectionFactory  election.Factory
+	ProcessorFactory process.Factory
+	Lifecycle        fx.Lifecycle
+}
+
+// NewManager creates a new namespace manager
+func NewManager(p ManagerParams) Manager {
+	manager := &namespaceManager{
+		cfg:              p.Cfg,
+		logger:           p.Logger.WithTags(tag.ComponentNamespaceManager),
+		electionFactory:  p.ElectionFactory,
+		processorFactory: p.ProcessorFactory,
+		namespaces:       make(map[string]*namespaceHandler),
+	}
+
+	p.Lifecycle.Append(fx.StartStopHook(manager.Start, manager.Stop))
+
+	return manager
+}
+
+// Start initializes the namespace manager and starts handling all namespaces
+func (m *namespaceManager) Start(ctx context.Context) error {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
```

Comment:
maybe use the `ctx` argument instead of a background context?

Reply:
This is a separate context.
Start will receive context from fx.Lifecycle that handles start timeout. 
This is a separate context to track the manager's stopping.